README.md
# koa-66

[![Node.js Version][node-image]][node-url]
[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Dependency Status][dep-image]][dep-url]
[![Coverage Status][cov-img]][cov-url]

Router middleware for [koa](https://github.com/koajs/koa/) v2.

> feedbacks are welcome

## Features

- [Plugin middleware support](https://github.com/menems/koa-66#plugin-support)
- Express like http verbs methods (including `all`)
- Express like use function
- Express like param function
- Automatic OPTIONS response
- Automatic HEAD when GET is present
- 501 and 405 status (throw capability with headers)
- Mount instance on specific path
- Multiple middleware as arguments
- Multiple middleware as array

## Installation

```bash
# with npm
$ npm install koa-66

$ with yarn
$ yarn add koa-66
```

## Example

```javascript
const Koa = require('koa');
const Router = require('koa-66');
const app = new Koa();

const router = new Router();
const mainRouter  = new Router();

router.param('id', (ctx, next, id) => {
        ctx.yolo = id;
        return next();
});

router.use(async function(ctx, next) {
    ctx.a = " ";
    await next();
});

router.get('/:id', (ctx, next) => {
    return next().then(() => {
        ctx.body += ctx.a + ctx.yolo;
   })
});

router.get('/:id', async function(ctx) {
    ctx.body = await Promise.resolve('hello');
});

mainRouter.mount('/pouet', router);

app.use(mainRouter.routes());

app.listen(1664);
// GET http://localhost:1664/pouet/world
// => hello world
```

## Example with __throw__ option

```javascript
const Koa = require('koa');
const Router = require('koa-66');
const app = new Koa();

const router = new Router();

app.use(async function(ctx, next) {
	try {
		await next();
	}catch(e){
		if (e.status === 405) {
			ctx.status = 405;
			ctx.set(e.headers);
		}
	}
})

router.get('/', (ctx) => ctx.body = 'hello');

app.use(router.routes({throw: true}));

app.listen(1664);

// > curl http://localhost:1664/ -I -X POST
//
// HTTP/1.1 405 Method Not Allowed
// allow: HEAD, GET
// Content-Type: text/plain; charset=utf-8
// Content-Length: 18
// Date: Wed, 04 Nov 2015 10:29:06 GMT
// Connection: keep-alive

```

## Plugin support

I don't know if Plugin is a good term for this feature.
The goal was to add cappability to register some middleware on a main Router that will be inject via config object on different route.
(ex: authentication  or acl behaviour).
Why? Because I am lazy to require some middleware in all my router script with generaly relatif path...

So I decided to add the possibility to inject an object at first parameter (that will be a config object) and adding an extra middleware that will be inject in middleware stack. To register this plugin just use a `plugin()`method.

```javascript
const Router = require('koa-66');
const main = new Router();

// you can use multiple middleware as arguments or array
main.plugin('authent', (ctx, next) => {
    // pick plugin config object on ctx.state.plugins.
    // here ctx.state.plugins.authent === true;
	// do stuff inject user on context for example
	return next();
	//or throw or do nothing that will stop execution of router stack
})

main.plugin('acl', (ctx, next) => {
    // here ctx.state.plugins.acl === ['admin'];
	// do stuff check role via ctx.state.plugins.acl for example
	return next();
	//or throw or do nothing that will stop execution of router stack
})

const router = new Router();

router.use({authent: true});
//options here is a boolean,
//but you can pass everything you want,
//and it will be inject as options

router.get('/private', {acl:['admin']}, 
	ctx => ctx.body = 'private'
)

main.mount('/api', router);
...	
// order of call /api/private
// 1 plugin authent
// 2 plugin acl
// 3 real middleware 

``` 

## Test
```bash
# npm test

```

[node-image]: https://img.shields.io/node/v/koa-66.svg?style=flat-square
[node-url]: https://nodejs.org
[npm-image]: https://img.shields.io/npm/v/koa-66.svg?style=flat-square
[npm-url]: https://npmjs.org/package/koa-66
[travis-image]: https://img.shields.io/travis/menems/koa-66/master.svg?style=flat-square
[travis-url]: https://travis-ci.org/menems/koa-66
[dep-image]: http://david-dm.org/menems/koa-66.svg?style=flat-square
[dep-url]:http://david-dm.org/menems/koa-66
[cov-img]:https://coveralls.io/repos/menems/koa-66/badge.svg?branch=master&service=github
[cov-url]:https://coveralls.io/github/menems/koa-66?branch=master


History.md

v1.0.1 / 2017-02-25
===================

  * doc: fix koa link
  * travis: add 7.6 support
  * core: koa 2.0.1 + yarn

v1.0.0 / 2016-06-22
===================

  * chore: update dependencies

v0.8.5 / 2016-05-27
===================

  * chore: add node6 test on travis

v0.8.4 / 2015-12-09
===================

  * Call param() middleware before the first actual match.

v0.8.3 / 2015-12-07
===================

  * fix: plugin need return

v0.8.2 / 2015-12-07
===================

  * fix: remove instanceof test on mount. (too many problem with npm link and different version)

v0.8.1 / 2015-12-07
===================

  * fix: plugins options are now in context

v0.8.0 / 2015-12-03
===================

  * chores: lint test
  * feat: remove del , no more reserved word
  * chores: lint
  * chore: typo

v0.7.3 / 2015-11-23
===================

  * chore: typo
  * attach the route to the context
  * chore : dependencies, readme

v0.7.2 / 2015-11-19
===================

  * test: add more test for no path and all method
  * chore: Readme update
  * feat: add multiple args on plugin
  * fix: bad test

v0.7.1 / 2015-11-14
===================

  * fix: resolve should have result...  #14

v0.7.0 / 2015-11-13
===================

  * feat: no path needed now, default to /
  * feat: add sanitizePath method
  * feat: add plugin support
  * refactor: remove unnecessary () and {}

v0.6.1 / 2015-11-11
===================

  * test: missing test
  * perf: remove unnecessary branch
  * chore : add istanbul, coverage

v0.6.0 / 2015-11-05
===================

  * chore: Readme update
  * refactor: remove const Koa66
  * fix: remove del from allow methods
  * feat: Support all() method for Router

v0.5.4 / 2015-11-04
===================

  * feat: Add throw option on routes();
  * chore: add dependencies badge + var -> const

v0.5.3 / 2015-11-02
===================

  * refactor: test with arrow function
  * chore: readme.md, package.json update

v0.5.2 / 2015-11-02
===================
  * wtf npm

v0.5.1 / 2015-11-02
===================

  * Fix test and add badges
  * dep: change to koa 2.0.0-alpha.2
  * Add travis
  * fix: readme

v0.5.0 / 2015-10-30
===================

  * Readme changed
  * fix: param() on children mount instance
  * Replace assert by throw
  * Add param() api
  * fix: no use of this in parseParams

v0.4.0 / 2015-10-29
===================

  * Add History.md
  * Add automatic options response
  * fix: methods on prototype instead in constructor
  * fix: multiple middleware on use functions
  * readme update

v0.3.0 / 2015-10-26
===================

  * refactor routes() and add HEAD request
  * add del to methods
  * don't catch decodeURIComponent error
  * Add 501 and 405

v0.2.2 / 2015-10-26
===================

  * fix readme

v0.2.1 / 2015-10-26
===================

  * More readme
  * remove unnecessary context param
  * Use instanceof instead of constructor.name
  * remove methods dependencies + refactor + multiple routes and middlewares

v0.1.3 / 2015-10-23
===================

  * fix: next call doesn't need ctx
  * Add sync test
  * Add install on readme
  * add babel as dependencies for test
  * Add eslint config
  * Add comments

v0.1.1 / 2015-10-22
===================

  * readme: add example
  * fix: use regexp
  * Initial commit



index.js
/**
 * Router for Koa v2
 *
 * @author blaz <blaz@menems.net>
 * @link https://github.com/menems/koa-66
 * @license MIT
 *
 */

'use strict';

const debug = require('debug')('koa-66');
const pathToRegexp = require('path-to-regexp');
const compose = require('koa-compose');
const util = require('util');

const methods = [
    'options',
    'head',
    'get',
    'post',
    'put',
    'patch',
    'delete'
];

/**
 * Expose Koa66 class.
 */
class Koa66 {

    /**
     * Initialise a new Koa66
     *
     * @api public
     */
    constructor() {
        this.stacks = [];
        this.methods = methods;
        this.plugs = [];
    }

    /**
     * Mount a Koa66 instance on a prefix path
     *
     * @todo: too memory expensive instance are clone
     * maybe add a keep args to remove the router instance mounted!
     *
     * @param  {string} prefix
     * @param  {Object} router  Koa66 instance
     * @return {Object}         Koa66 instance
     * @api public
     */
    mount(prefix, router) {
        if (typeof router != 'object')
            throw new TypeError('require a Koa66 instance');

        router.stacks.forEach(s => {
            if (s.paramKey)
                this.register(s.methods, prefix + s.path, s.paramKey, s.middleware);
            else
                this.register(s.methods, prefix + s.path, s.middleware);
        });
        return this;
    }

    /**
     * Use given middleware before route callback
     *
     * @param  {String|Function} path
     * @param  {Function} fn
     * @return {Object} Koa66 instance
     * @api public
     */
    use() {
        const args = Array.prototype.slice.call(arguments);
        if (typeof args[0] != 'string')
            args.unshift('(.*)');
        args.unshift(false);
        return this.register.apply(this, args);
    }

    /**
     * param express like function
     *
     * @param  {String}   key
     * @param  {Function} fn
     * @return {Object}
     */
    param(key, fn) {
        if (typeof key != 'string' || typeof fn != 'function')
            throw new TypeError('usage: param(string, function)');
        return this.register(false, '(.*)', key, fn);
    }

    /**
     * plugin registration method
     *
     * @param  {String}   name
     * @param  {[Function]...}
     * @return {Object}
     */
    plugin(name) {
        if (typeof name != 'string')
            throw new TypeError('usage: plugin(string, function)');

        const args = Array.prototype.slice.call(arguments, 1);

        let middlewares = [];
        args.forEach(m => {
            if (util.isArray(m)) middlewares = middlewares.concat(m);
            if (typeof m == 'function') middlewares.push(m);
        });
        this.plugs[name] = compose(middlewares);
        return this;
    }

    /**
     * Expose middleware for koa
     *
     * @return {Function}
     * @api public
     */
    routes(options) {
        options = options || {};

        return (ctx, next) => {
            const middlewares = [];
            const allowed = [];
            const paramMiddlewares = [];
            let matched;

            this.stacks.forEach(route => {
                // path test
                if (!route.regexp.test(ctx.path)) return;

                // Save the route so we can access ctx.route.path
                ctx.route = route;

                if (route.paramNames)
                    ctx.params = this.parseParams(ctx.params, route.paramNames, ctx.path.match(route.regexp).slice(1));

                if (route.paramKey) {
                    paramMiddlewares[route.paramKey] = (ctx, next) =>
                        route.middleware(ctx, next, ctx.params[route.paramKey]);
                    return;
                }

                if ( typeof route.middleware === 'object') {
                    for (let i in route.middleware) {
                        if ( this.plugs[i]) {
                            middlewares.push((ctx, next) => {
                                ctx.state.plugins = ctx.state.plugins || {};
                                ctx.state.plugins[i] = route.middleware[i];
                                return this.plugs[i](ctx, next);
                            });
                        }
                    }
                    return;
                }

                if (!route.methods)
                    return middlewares.push(route.middleware);

                if (route.methods.indexOf('GET') !== -1)
                    allowed.push('HEAD');
                route.methods.forEach(m => allowed.push(m));

                // method test
                if ((route.methods.indexOf(ctx.method) !== -1) ||
                    (ctx.method === 'HEAD' && route.methods.indexOf('GET') !== -1)) {
                    matched = true;
                    for (let i in ctx.params) {
                        if (paramMiddlewares[i])
                            middlewares.push(paramMiddlewares[i]);
                            delete paramMiddlewares[i]
                    }
                    middlewares.push(route.middleware);
                }
            });

            // only use middleware
            if (!allowed.length) return next();

            // 501
            if (this.methods.indexOf(ctx.method.toLowerCase()) === -1) {
                if (options.throw)
                    ctx.throw(501);
                ctx.status = 501;
                return next();
            }

            // 405
            if (!matched) {
                // automatic OPTIONS response
                if (ctx.method === 'OPTIONS') {
                    ctx.status = 204;
                    return next();
                }

                const _allowed = allowed.filter((value, index, self) => self.indexOf(value) === index).join(', ');

                if (options.throw)
                    ctx.throw(405, {headers: {allow: _allowed}});
                ctx.status = 405;
                ctx.set('Allow', _allowed);
                return next();
            }

            return compose(middlewares)(ctx, next);
        };
    }

    /**
     * Register a route for all methods.
     *
     * @param  {String}  path
     * @return {[Functions]...}  middleware
     */
    all() {
        const args = Array.prototype.slice.call(arguments);

        if (typeof args[0] != 'string') args.unshift('/');

        args.unshift(methods);
        return this.register.apply(this, args);
    }

    sanitizePath(path) {
        if (!path) return '/';

        return '/' + path
            .replace(/^\/+/i, '')
            .replace(/\/+$/, '')
            .replace(/\/{2,}/, '/');
    }


    /**
     * Register a new middleware, http route or use middeware
     *
     * @param  {string} methods
     * @param  {string} path
     * @param  {Function} middleware
     * @return {Object} Koa66 instance
     * @api private
     */
    register(methods, path) {
        debug('Register %s %s', methods, path);
        let middlewares;
        let paramKey;
        if (typeof arguments[2] === 'string') {
            paramKey = arguments[2];
            middlewares = Array.prototype.slice.call(arguments, 3);
        } else
            middlewares = Array.prototype.slice.call(arguments, 2);

        if (!middlewares.length)
            throw new Error('middleware is required');

        middlewares.forEach(m => {
            if (Array.isArray(m)) {
                m.forEach(_m => this.register(methods, path, _m));
                return this;
            }

            if ( typeof m !== 'function' && typeof m !== 'object')
                throw new TypeError('middleware must be a function');

            const keys = [];
            path = (!path || path === '(.*)' || util.isRegExp(path)) ? path : this.sanitizePath(path);
            const regexp = pathToRegexp(path, keys);

            const route = {
                path: path,
                middleware: m,
                regexp: regexp,
                paramNames: keys
            };

            if (paramKey) route.paramKey = paramKey;

            if (methods) route.methods = methods.map((m) => m.toUpperCase());

            this.stacks.push(route);
        });
        return this;
    }

    /**
     * parse params from route
     *
     * @param  {[Object]} paramNames
     * @param  {[String]} captures
     * @return {[Object]}
     * @api private
     */
    parseParams(params, paramNames, captures) {
        const len = captures.length;
        params = params || {};

        for (let i = 0; i < len; i++) {
            if (paramNames[i]) {
                let c = captures[i];
                params[paramNames[i].name] = c ? decodeURIComponent(c) : c;
            }
        }
        return params;
    }
}

/**
 * Add http methods
 *
 * @param  {String}  path
 * @return {[Functions]...}  middleware
 */
methods.forEach(method => {
    Koa66.prototype[method] = function() {
        const args = Array.prototype.slice.call(arguments);

        if (typeof args[0] !== 'string')
            args.unshift('/');

        args.unshift([method]);
        return this.register.apply(this, args);
    };
});

module.exports = Koa66;
